### PR TITLE
Fix gles3 draw

### PR DESCRIPTION
**Pull request by Kangz**
PTAL, this fixes gles3/draw.html

### CODE REVIEW

**Comment by Kangz** [comment]
cc @kenrussell @zhenyao

**Comment by Kangz** [comment]
Ah, it also makes the random draw tests harder to pass (it draws more geometry). That's a side effect of removing the workarounds for the reference rasterizer only supporting quads

File: sdk/tests/deqp/framework/referencerenderer/rrVertexAttrib.js [hunk 1]
```diff
@@ -220,6 +220,19 @@ var rrGenericVector = framework.referencerenderer.rrGenericVector;
         DE_ASSERT(rrVertexAttrib.isValidVertexAttrib(vertexAttrib));
         /** @type {goog.NumberArray} */ var dst;
 
+        var arrayType = null;
```
Comment (Kangz) [inline sdk/tests/deqp/framework/referencerenderer/rrVertexAttrib.js]: The C++ code is different from the original, and modified version and use templating.

File: sdk/tests/deqp/functional/gles3/es3fDrawTests.js [hunk 2]
```diff
@@ -88,15 +88,14 @@ goog.scope(function() {
     es3fDrawTests.addTestIterations = function(test, baseSpec, type) {
         var spec = /** @type {glsDrawTests.DrawTestSpec} */ (deUtil.clone(baseSpec));
```
Comment (Kangz) [inline sdk/tests/deqp/functional/gles3/es3fDrawTests.js]: Changes to this file make stuff closer to the C++ version

File: sdk/tests/deqp/modules/shared/glsDrawTests.js [hunk 3]
```diff
@@ -1443,59 +1443,7 @@ goog.scope(function() {
...
-                    components[componentNdx] = glsDrawTests.GLValue.getRandom(vertexParameters.rnd, vertexParameters.min, vertexParameters.max);
-                }
-            }
-        }
-        return components;
-    };
-
-    /**
-     * createBasicArray - Modified JS version to ensure quad generation for compatibility with refrast.
+     * createBasicArray
```
Comment (Kangz) [inline sdk/tests/deqp/modules/shared/glsDrawTests.js]: Change in this file make the code massively closer to the C++ version, some small differences are noted below

File: sdk/tests/deqp/modules/shared/glsDrawTests.js [hunk 4]
```diff
@@ -1524,228 +1475,56 @@ goog.scope(function() {
...
-                        vertex4[oppositeaxis] = vertex2[oppositeaxis];
-                        vertex4[parallelaxis] = vertex3[parallelaxis];
-
-                        quadVertices = [vertex1, vertex2, vertex3, vertex4];
+            var components = [];
+
+            for (var componentNdx = 0; componentNdx < componentCount; componentNdx++) {
+                var getRandomComponent = function() {
+                    // For packed formats we can't use GLValue
+                    if (packed) {
```
Comment (Kangz) [inline sdk/tests/deqp/modules/shared/glsDrawTests.js]: The C++ version use templating on the vertex attribute type, while here we use GLValue that doesn't handle the (U)INT_2_10_10_10_REV formats so we special case these formats with the packed variable like the previous code was doing in various places.

File: sdk/tests/deqp/modules/shared/glsDrawTests.js [hunk 5]
```diff
@@ -1524,228 +1475,56 @@ goog.scope(function() {
...
-                }
-                case glsDrawTests.DrawTestSpec.Primitive.TRIANGLE_FAN: {
-                    if (vertexNdx < first + 4) {
-                        vertex1 = generateVertex(components, vertexParameters);
+                };
 
-                        parallelaxis = rnd.getBool() ? 0 : 1;
-                        oppositeaxis = parallelaxis ? 0 : 1;
+                var component = getRandomComponent();
+                var componentFloat = (component instanceof glsDrawTests.GLValue) ? component.toFloat() : component;
```
Comment (Kangz) [inline sdk/tests/deqp/modules/shared/glsDrawTests.js]: The dEQP code doesn't use componentFloat and uses multiple loops on componentNdx. The logic here is he same but expressed slightly differently.

**Comment by kenrussell** [comment]
Awesome. LGTM from my standpoint. @zhenyao has been reviewing all of the dEQP fixes so let's let him take a look and merge.

File: sdk/tests/deqp/functional/gles3/es3fDrawTests.js [hunk 6]
```diff
@@ -109,10 +108,10 @@ goog.scope(function() {
             test.addIteration(spec, 'instance count = ' + spec.instanceCount);
         } else if (type == es3fDrawTests.TestIterationType.INDEX_RANGE) {
             spec.indexMin = 0;
-            spec.indexMax = 24;
+            spec.indexMax = 23;
             test.addIteration(spec, 'index range = [' + spec.indexMin + ', ' + spec.indexMax + ']');
 
-            spec.indexMin = 24;
+            spec.indexMin = 23;
             spec.indexMax = 41;
```
Comment (zhenyao) [inline sdk/tests/deqp/functional/gles3/es3fDrawTests.js]: c++ code this is 40. Why align all other numbers but not this one?

**Comment by zhenyao** [comment]
This is great work.  Mostly looks good with one question.

File: sdk/tests/deqp/functional/gles3/es3fDrawTests.js (hunk 6 above)
Comment (Kangz) [inline sdk/tests/deqp/functional/gles3/es3fDrawTests.js]: I didn't realized this one had changed, will align with the C++ code.

Comment (Kangz) [inline sdk/tests/deqp/functional/gles3/es3fDrawTests.js]: Done.

**Comment by zhenyao** [comment]
LGTM
Thank you!

**Pull request merged**